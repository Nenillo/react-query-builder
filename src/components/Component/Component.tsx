import React from 'react';
import styled from 'styled-components';
import { colors } from '../../constants/colors';

const StyledComponent = styled.div`
  display: grid;
  grid-auto-flow: column;
  padding: 0.7rem;
  margin: 0.5rem 0;
  border: 1px solid ${colors.medium};
  background-color: #fff;
`;

const Content = styled.div`
  display: grid;
  grid-auto-columns: min-content;
  grid-auto-flow: column;
  grid-gap: 0.5rem;
`;

const Controls = styled.div`
  display: grid;
  grid-auto-flow: column;
  grid-auto-columns: min-content;
  grid-gap: 0.5rem;
  justify-self: flex-end;
`;

export interface ComponentProps {
  children: React.ReactNode | React.ReactNodeArray;
  controls: React.ReactNode | React.ReactNodeArray;
  className?: string;
}

export const Component: React.FC<ComponentProps> = ({
  children,
  controls,
  className,
}) => (
  <StyledComponent className={className}>
    <Content>{children}</Content>
    <Controls>{controls}</Controls>
  </StyledComponent>
);
