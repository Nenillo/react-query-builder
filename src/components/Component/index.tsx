import React, { useContext } from 'react';
import styled from 'styled-components';
import { BuilderFieldOperator, BuilderContext } from '../Builder';
import { FieldSelect } from '../Widgets/FieldSelect';
import { Boolean } from '../Widgets/Boolean';
import { Select } from '../Widgets/Select';
import { SelectMulti } from '../Widgets/SelectMulti';
import { OperatorSelect } from '../Widgets/OperatorSelect';
import { Input } from '../Widgets/Input';
import {
  isBoolean,
  isReactText,
  isReactTextArray,
  isOptionList,
} from '../../utils/types';
import { Component as ComponentContainer } from './Component';

const BooleanContainer = styled.div`
  align-self: center;
`;

export interface ComponentProps {
  field: string;
  value?: React.ReactText | React.ReactText[] | boolean;
  operator?: BuilderFieldOperator;
  id: string;
}

export const Component: React.FC<ComponentProps> = ({
  field: fieldRef,
  value: selectedValue,
  operator,
  id,
}) => {
  const { fields } = useContext(BuilderContext);

  if (fields) {
    if (fieldRef === '') {
      return (
        <ComponentContainer>
          <FieldSelect selectedValue={''} id={id} />
        </ComponentContainer>
      );
    } else {
      try {
        const fieldIndex = fields.findIndex(item => item.field === fieldRef);
        const { field, operators, type, value: fieldValue } = fields[
          fieldIndex
        ];

        const operatorsOptionList =
          operators && operators.map(item => ({ value: item, label: item }));

        return (
          <ComponentContainer>
            <FieldSelect selectedValue={field} id={id} />

            {type === 'BOOLEAN' && isBoolean(selectedValue) && (
              <BooleanContainer>
                <Boolean id={id} selectedValue={selectedValue} />
              </BooleanContainer>
            )}

            {type === 'LIST' &&
              isReactText(selectedValue) &&
              isOptionList(fieldValue) &&
              isOptionList(operatorsOptionList) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Select
                    id={id}
                    selectedValue={selectedValue}
                    values={fieldValue}
                  />
                </>
              )}

            {type === 'MULTI_LIST' &&
              isOptionList(fieldValue) &&
              isOptionList(operatorsOptionList) &&
              isReactTextArray(selectedValue) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <SelectMulti
                    id={id}
                    values={fieldValue}
                    selectedValue={selectedValue}
                  />
                </>
              )}

            {type === 'TEXT' &&
              isOptionList(operatorsOptionList) &&
              (isReactText(selectedValue) ||
                isReactTextArray(selectedValue)) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Input type="text" value={selectedValue} id={id} />
                </>
              )}

            {type === 'NUMBER' &&
              isOptionList(operatorsOptionList) &&
              (isReactText(selectedValue) ||
                isReactTextArray(selectedValue)) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Input type="number" value={selectedValue} id={id} />
                </>
              )}

            {type === 'DATE' &&
              isOptionList(operatorsOptionList) &&
              (isReactText(selectedValue) ||
                isReactTextArray(selectedValue)) && (
                <>
                  <OperatorSelect
                    id={id}
                    values={operatorsOptionList}
                    selectedValue={operator}
                  />
                  <Input type="date" value={selectedValue} id={id} />
                </>
              )}
          </ComponentContainer>
        );
      } catch (e) {
        console.error(`Field "${fieldRef}" not found in fields definition.`);
      }
    }
  }

  return null;
};
