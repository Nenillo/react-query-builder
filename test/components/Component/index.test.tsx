import { Component } from '../../../src/components/Component/index';
import { shallow, mount } from 'enzyme';
import {
  BuilderContext,
  BuilderFieldProps,
  BuilderComponentsProps,
  defaultComponents,
} from '../../../src/components/Builder';
import { strings } from '../../../src/constants/strings';
import React from 'react';

const components: BuilderComponentsProps = defaultComponents;
const fields: BuilderFieldProps[] = [
  {
    field: 'MOCK_FIELD_1',
    label: 'Mock Field',
    operators: ['EQUAL'],
    type: 'BOOLEAN',
  },
  {
    field: 'MOCK_FIELD_2',
    label: 'Mock Field 2',
    operators: ['EQUAL'],
    type: 'TEXT',
  },
  {
    field: 'MOCK_FIELD_3',
    label: 'Mock Field 3',
    operators: ['EQUAL'],
    type: 'DATE',
  },
  {
    field: 'MOCK_FIELD_4',
    label: 'Mock Field 4',
    operators: ['EQUAL'],
    type: 'NUMBER',
  },
  {
    field: 'MOCK_FIELD_5',
    label: 'Mock Field 5',
    operators: ['EQUAL'],
    type: 'STATEMENT',
  },
  {
    field: 'MOCK_FIELD_6',
    label: 'Mock Field 6',
    operators: ['EQUAL'],
    type: 'LIST',
  },
  {
    field: 'MOCK_FIELD_7',
    label: 'Mock Field 7',
    operators: ['EQUAL'],
    type: 'MULTI_LIST',
  },
];
const data: any[] = [
  {
    type: 'GROUP',
    value: 'AND',
    id: 'test-1',
    isNegated: false,
    children: ['test-2'],
  },
  {
    field: 'MOCK_FIELD_1',
    value: true,
    id: 'test-2',
    parent: 'test-1',
  },
  {
    field: 'MOCK_FIELD_2',
    value: '',
    id: 'test-3',
    parent: 'test-1',
  },
  {
    field: 'MOCK_FIELD_3',
    value: '',
    id: 'test-4',
    parent: 'test-1',
  },
  {
    field: 'MOCK_FIELD_4',
    value: '',
    id: 'test-5',
    parent: 'test-1',
  },
  {
    field: 'MOCK_FIELD_5',
    value: '',
    id: 'test-6',
    parent: 'test-1',
  },
  {
    field: 'MOCK_FIELD_6',
    value: '',
    id: 'test-7',
    parent: 'test-1',
  },
  {
    field: 'MOCK_FIELD_7',
    value: '',
    id: 'test-8',
    parent: 'test-1',
  },
];
const setData = jest.fn();
const onChange = () => jest.fn();

describe('#components/Component', () => {
  it('Tests snapshot', () => {
    expect(
      shallow(
        <BuilderContext.Provider
          value={{ components, fields, data, strings, setData, onChange }}
        >
          <Component id="test-2" field="MOCK_FIELD_1" />
        </BuilderContext.Provider>
      )
    ).toMatchSnapshot();
  });

  it('Tests full behavior', () => {
    const wrapper = mount(
      <BuilderContext.Provider
        value={{ components, fields, data, strings, setData, onChange }}
      >
        <Component data-test="Component[0]" id="test-2" field="" />
        <Component data-test="Component[1]" id="test-2" field="MOCK_FIELD_1" />
        <Component data-test="Component[2]" id="test-3" field="MOCK_FIELD_2" />
        <Component data-test="Component[3]" id="test-4" field="MOCK_FIELD_3" />
        <Component data-test="Component[4]" id="test-5" field="MOCK_FIELD_4" />
        <Component data-test="Component[5]" id="test-6" field="MOCK_FIELD_5" />
        <Component data-test="Component[6]" id="test-7" field="MOCK_FIELD_6" />
        <Component data-test="Component[7]" id="test-8" field="MOCK_FIELD_7" />
        <Component data-test="Component[8]" id="test-9" field="SOME_FIELD_THAT_DOESNT_EXISTS" />
      </BuilderContext.Provider>
    )

    expect(wrapper.find('[data-test="Component[0]"]')).toBeDefined()
    expect(wrapper.find('[data-test="Component[1]"]')).toBeDefined()
    expect(wrapper.find('[data-test="Component[2]"]')).toBeDefined()
    expect(wrapper.find('[data-test="Component[3]"]')).toBeDefined()
    expect(wrapper.find('[data-test="Component[4]"]')).toBeDefined()
    expect(wrapper.find('[data-test="Component[5]"]')).toBeDefined()
    expect(wrapper.find('[data-test="Component[6]"]')).toBeDefined()
    expect(wrapper.find('[data-test="Component[7]"]')).toBeDefined()
    expect(Object.keys(wrapper.find('[data-test="Component[8]"]')).length).toBe(0)

    wrapper.find('[data-test="Component[1]"] button').simulate('click')

    expect(Object.keys(wrapper.find('[data-test="Component[1]"]')).length).toBe(0)
  })

  it('Tests no srtrings scenario', () => {
    const wrapper = mount(
      <BuilderContext.Provider
        value={{ components, fields, data, strings: {}, setData, onChange }}
      >
        <Component id="test-9" field="" />
      </BuilderContext.Provider>
    );

    expect(Object.keys(wrapper).length).toBe(0);
  });
});
