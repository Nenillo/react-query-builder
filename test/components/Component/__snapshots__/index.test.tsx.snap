// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#components/Component Tests snapshot 1`] = `
<Component
  field="MOCK_FIELD_1"
  id="test-2"
/>
`;

exports[`#components/Component Tests snapshot 2`] = `
<Component
  field="MOCK_FIELD_1"
  id="test-2"
/>
`;
